### DISCUSSION TITLE
Why Python?

### DISCUSSION THREAD

**Discussion by maaft** [geldata/gel#3442] [General]
I was wondering, why did you choose python as the main language for EdgeDB? Why not languages that are "made" for such system-level tasks like `go`, `rust`, ... ?

I'm honestly curious as from my experience python has a few issues (99% chance I'm wrong though):
- although types can be enforced by IDE/CI, still many 3rd party libs don't bother to type their stuff. Programming without proper typing is hard and error-prone
- bad multi-processing experience (not sure about asyncio though, never used it)
- script language (but I assume you use cython where speed matters)
- huge deployment sizes (can't be compiled natively for any platform, so python interpreters have to be shipped)

Anyway, I'm eager to try out edgedb and have my prejudices busted! :)

**Comment by msullivan** [comment]
There are a bunch of parts to this answer, and I'm just going to address the ones that are most relevant to the parts that I work on:

We heavily use mypy for typing, and we don't use a lot of 3rd party libraries so we haven't run into many problems with them being untyped. (And in general most of the libraries we use don't have a very big surface in edgedb where we use them? And lots of libraries *do* have typing information now.)

The largest components of edgedb (by LOC, not by execution time) are the EdgeQL -> postgres SQL compiler and the engine for generation migrations. mypy-typed Python is actually a very productive language for writing compilers in! I would choose it over go for that task ten times out of ten. Rust can handle this sort of compiler building task pretty well, but you still wind up needing to devote a lot of time to thinking about memory management related issues in a situation where in my opinion you don't really care about them. Personally ocaml would probably be my first pick for writing this sort of compiler in, but choosing such a niche language would have plenty of its own downsides!

The compiler is definitely slower in python than it could be in other languages (though there are some avenues for speeding it up!), but that's why we have a compiler query cache :)

Multiprocessing winds up working pretty well for us because we have a bunch of compiler tasks that really are totally independent.

**Reply by maaft** [reply]
Okay, thank you so far. I've never written compilers, so thank you for this insight that python seems to be a good choice for that.

What about integration with other projects / ease of deployment?

For integration, I mainly ask because I'm unsure how I can add my business logic (query and mutation hooks; in my language of choice) to edgedb. But I'm not even halfway through the tutorial, so maybe this will get answered soon ;)

**Reply by fantix** [reply]
> For integration, I mainly ask because I'm unsure how I can add my business logic (query and mutation hooks; in my language of choice) to edgedb.

You can use one of our EdgeDB client libraries - currently supporting JavaScript/TypeScript, Python, Go and Deno - in your application and integrate EdgeDB though TCP connections, or enable the [EdgeQL over HTTP extension](https://www.edgedb.com/docs/clients/90_edgeql/index) and issue HTTP requests directly from your application to the EdgeDB server. 

I hope I didn't get you wrong, but in case you meant hacking into EdgeDB Server itself, yeah you'll have to use Python, Cython and Rust. (Did I mention WebAssembly? SPOILER ALERT)

> huge deployment sizes (can't be compiled natively for any platform, so python interpreters have to be shipped)

Your application doesn't usually ship with a bundled EdgeDB release - for example you write a Discord robot in Golang, you deploy your Go server separately from running an EdgeDB instance.

For EdgeDB Server itself, yes it's coming with a Python interpreter, and we set up all the CD automation and built the EdgeDB CLI to make all that into just one command like `edgedb project init` or `edgedb instance create` in development environments. This would be consistently the same convenient when we have EdgeDB cloud but only with less local downloads and installation.

**Reply by maaft** [reply]
> I hope I didn't get you wrong, but in case you meant hacking into EdgeDB Server itself, yeah you'll have to use Python, Cython and Rust. (Did I mention WebAssembly? SPOILER ALERT)

Yes, that is something that I meant. The reason is that real-life applications almost always need to handle some kind of business logic, which might be arbitrarily complex and in any language out there. This is also why I started this feature request: #3455 

> Your application doesn't usually ship with a bundled EdgeDB release - for example you write a Discord robot in Golang, you deploy your Go server separately from running an EdgeDB instance

There is a use-case for offline-first apps: Consider the same web-app, where you usually do your work offline (e.g. due to security restrictions inside a fab, or missing internet connectivity in rural areas). As soon as you have internet, your offline-database can sync with the main cloud-database and you can use online-services (like long running ML tasks). This use-case would greatly benefit from being able to ship the database with your application as you don't have to write 2 different GUIs.

Anyway, I think I got the answers I was looking for. Thanks a lot! And if you want, feel free to participate in #3455 :)

**Reply by CanRau** [reply]
Came here looking for exactly the offline-first use case @maaft is talking about 😅
Was excited to stumble upon a new DB, as the embeddable & scalable (on server), ideally with sync, market is still way to small 🥲
